File: Library.m
#import "Library.h"
#import "Book.h"

@interface Library()

@property (nonatomic, retain, readwrite) NSMutableArray *books;
@property (nonatomic, retain, readwrite) NSMutableArray *libraryUrls;  //all the floders where stories are stored
@property (nonatomic, retain, readwrite) NSString *name;

@end

@implementation Library

-(id)init{
    return nil;
}

-(BOOL) deleteBook: (Book*) delBook {
    NSError* err = nil;
    BOOL result = [[NSFileManager defaultManager] removeItemAtURL:delBook.bookUrl error:&err];
    
    [self getBooks];
    
    return result;
}

-(id) initWithLibraryFolderUrl:(NSURL*) LibraryFolderURL andName: (NSString*) name {
    
    if (self = [super init]) {
        self.books = [[NSMutableArray alloc] init];
        [self addLibraryUrl: LibraryFolderURL];
        self.name = name;
    }
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(getBooks)
                                                 name:@"DownloadFinished"
                                               object:nil];
    return self;
}

-(int)getBooks{
    //get all the books that are in the libraries folder
    //put those books in the books array
    NSArray *bookFolders;
    NSFileManager *fm = [NSFileManager defaultManager];
    
    [self.books removeAllObjects];
    
    for (NSURL* libURL in self.libraryUrls) {
        bookFolders = [fm contentsOfDirectoryAtURL:libURL includingPropertiesForKeys:nil options:NSDirectoryEnumerationSkipsHiddenFiles error:nil];
        for (NSURL *bookFolderUrl in bookFolders) {
            
            Book *someBook = [[Book alloc] initWithBookFolderURL:bookFolderUrl];
            if (someBook) {
                [self.books addObject:someBook];
            }
        }
    }
    
    [[NSNotificationCenter defaultCenter] postNotificationName:@"LibraryChanged" object:nil];
    return 0;
}

-(BOOL)addLibraryUrl: (NSURL*)libraryURL{
    
    if (!self.libraryUrls) {
        self.libraryUrls = [[NSMutableArray alloc]init];
    }
    
    [self.libraryUrls addObject:libraryURL];
    [self getBooks];
    return TRUE;
}

@end

File: Library.swift
import Foundation

extension Notification.Name {
    static let libraryChanged = Notification.Name("LibraryChanged")
}

class Library: NSObject {
    private var libraryURLs: [URL] = []
    @objc var title: String = ""
    @objc var books: [Book] = []

    @objc init(WithFolder folder: URL, AndTitle title: String) {
        self.title = title
        self.libraryURLs.append(folder)
        super.init()
        self.books = loadBooksFrom(url: folder)
    }

    private func loadBooksFrom(urls: [URL]) -> [Book] {
        var books: [Book] = []
        for url in urls {
            books.append(contentsOf: loadBooksFrom(url: url))
        }
        return books
    }

    private func loadBooksFrom(url: URL) -> [Book] {
        var books: [Book] = []
        do {
            let bookFolders = try FileManager.default.contentsOfDirectory(at: url, includingPropertiesForKeys: nil, options: .skipsHiddenFiles)
            for url in bookFolders {
                if let book = Book(sourceFolder: url) {
                    books.append(book)
                }
            }
        } catch {
            print("could not load directory contents because of: \(error)")
        }
        return books
    }

    @objc func reloadLibrary() {
        self.books = loadBooksFrom(urls: self.libraryURLs)
        NotificationCenter.default.post(name: Notification.Name.libraryChanged, object: self)
    }
}
